src/AppInstallerCLICore/Core.cpp

```diff
@@ -41,6 +42,63 @@ namespace AppInstaller::CLI
         };
     }
 
+    void  OnProgress(uint32_t reportType, uint64_t current, uint64_t maximum, uint32_t progressType, uint32_t executionPhase)
+    {
+        // Rest of this method is just Local Testing and Demo
+        std::cout << "Execution Stage:";
+        std::cout << executionPhase;
+        std::cout << "   reportType: ";
+        std::cout << reportType;
+        std::cout << "   current: ";
+        std::cout << current;
+        std::cout << "   maximum: ";
+        std::cout << maximum;
+        std::cout << "   ProgressType : ";
+        std::cout << (uint32_t)progressType;
+        std::cout << "\n";
+
+        if (current == maximum)
+        {
+            // Current Execution Phase is Complete
+            std::cout << "Current Execution Phase is now complete\n";
+        }
+    }
+
+    void TestCOMScenario()
+    {
+        init_apartment();
+
+        // proposal: BEGIN what COM Interface should do
+        // Eventually, COM Interface has to be aware whether the caller wants CLI ouptut or not. And appropriate set null streams or std::cout/cin streams
+        class NullStreamBuf : public std::streambuf {};
+        NullStreamBuf nullStreamBuf;
+        std::ostream nullOstream(&nullStreamBuf);
+        std::istream nullIstream(&nullStreamBuf);
+
+        AppInstaller::COMContext context(nullOstream, nullIstream);
+        context.SetProgressCallbackFunction(OnProgress);
+
+        std::vector<std::string> utf8Args;
+        utf8Args.emplace_back(Utility::ConvertToUTF8(L"install"));
+        utf8Args.emplace_back(Utility::ConvertToUTF8(L"mozilla.firefox"));
+        // proposal: END what COM Interface should do
+
+        // proposal: BEGIN what COMContext could do in it's own new method to not bother COMInterface with below
+        Invocation invocation{ std::move(utf8Args) };
+
+        std::unique_ptr<Command> command = std::make_unique<RootCommand>();
+        std::unique_ptr<Command> subCommand = command->FindSubCommand(invocation);
+        while (subCommand)
+        {
+            command = std::move(subCommand);
+            subCommand = command->FindSubCommand(invocation);
+        }
+
+        command->ParseArguments(invocation, context.Args);
+        command->Execute(context);
```

Comment:
The COM function should do all of this, as:
```
context.Args.Add(Id, "firefox");
InstallCommand command(RootCommand::Name());
command.Execute(context);
```
There is less burden placed on the COM function to know the names of the arguments than to know how to construct a command line for them.

Reply:
std::string str = "mozilla.firefox";
        context.Args.AddArg(Execution::Args::Type::Query, str);
        RootCommand rootCommand;
        InstallCommand command(rootCommand.Name());
        command.Execute(context);

Reply:
Yes-ish.  I believe that our expectation around the COM API is that install would not be a search-like install as we have for the command line but an action taken on a well known value.  So that is why I would say to pass Id and Exact args, not Query.

As for the command construction, I would personally put them behind an enum driven CreateCommand method to centrally locate that knowledge on constructing the proper parentage.

Reply:
Actually, I wanted to get clarification in that regards. my guess was that the source will only have valid full values to pass to winget core and no search would be required. So, thanks for clarifying without even me asking. I stumbled upon this issue as I as was trying to initially test for just firefox and found that the same input for winget.exe is giving me multiple matches found error where as my test gave me no match found error and figured it's the difference between using Id and Query.

Do you suggest the changes for enum drive CreateComand to be part of this PR ?

Reply:
I would expect that to be part of the COM object helpers, so not necessary in this PR.